design/multiple-listeners-design.md

```diff
@@ -0,0 +1,147 @@
+- Envoy RouteConfigurations will be generated for each Envoy Listener. The naming scheme will be described in the Detailed Design section below. 
+
+For Gateway mode, we must also consider what happens when HTTPProxy or Ingress is used for routing configuration.
+The details of this use case are described in the Compatibility section below.
+
+## Detailed Design
+
+### internal/gatewayapi
+The `ValidateListeners` function logic must change, to allow any number of unique ports, while still validating protocols and hostnames for each port.
+Instead of returning a single insecure and secure port number, it will return a slice of ports.
+
+### internal/provisioner
+The Gateway controller will use the results of the `ValidateListeners` function (above) to add all relevant ports to `contourModel.Spec.NetworkPublishing.Envoy.Ports`.
+These ports will be populated on the Envoy service and deployment/daemonset.
+
+Currently, the single insecure port is always mapped to port 8080 inside the Envoy container, and the single secure port is always mapped to 8443.
+Supporting any number of Listeners requires a more flexible port mapping scheme, since the requested Gateway Listener ports could be privileged ports (e.g. 1-1023), non-privileged ports (e.g. 8080, 8443), and/or node ports (e.g. 32080).
+The main constraint is that, since Envoy runs as non-root, it cannot bind to privileged ports within the container.
+As such, any requested privileged Gateway Listener port must be mapped to a non-privileged port inside the Envoy container.
+The scheme proposed is as follows:
+- any privileged Gateway Listener port (1-1023) is mapped to the range 64513-65535 within the container.
+- any other Gateway Listener port (1024-65535) is mapped as-is within the container.
+- conflicts between mapped privileged ports, and high non-privileged ports (in the range 64513-65535), will be detected and reported as status conditions on the Gateway/Listener, resulting in those Listeners not being programmed.
+
+### internal/dag
+The first step in DAG processing will now be to populate Listeners, based on either the Gateway spec (for Gateway mode), or the CLI flags/config file/ContourConfiguration (for non-Gateway mode).
+The DAG will now store desired Listener names, addresses and "inside" ports.
+Subsequently, in the Ingress/HTTPProxy/Gateway API processors, virtual hosts will be added directly to the appropriate DAG Listener(s), rather than to the root of the DAG.
+The final step of DAG processing will still prune and sort Listeners and virtual hosts.
+
+The Listener naming scheme will be as follows:
+- for non-Gateway mode, the default Listeners will continue to be named `ingress_http` and `ingress_https` for compatibility.
+- for Gateway mode, Listeners will be named `<http|https>_<port number>`, e.g. `http_80` or `https_8443`.
+
+### internal/xdscache
+All DAG Listeners will be converted to Envoy Listeners, using the DAG Listeners' name, address and port.
```

Comment:
might want to mention the stat prefix for the new Gateway mode Listeners here too, in case there are any changes